core/src/main/java/org/elasticsearch/common/xcontent/NamedXContentRegistry.java

```diff
@@ -0,0 +1,150 @@
+import org.elasticsearch.common.ParseFieldMatcherSupplier;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.xcontent.support.DelegatingXContentParser;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static java.util.Collections.unmodifiableMap;
+
+/**
+ * Registry of "named" XContent parsers that can read using {@link XContentParser#namedObject(Class, String, ParseFieldMatcherSupplier)}.
+ */
+public class NamedXContentRegistry {
+    /**
+     * Parses an object with the type {@code T} from parser.
+     */
+    public interface FromXContent<T, C> {
+        T fromXContent(XContentParser parser, C context) throws IOException;
+    }
+
+    /**
+     * Entry that will be added to the registry.
+     */
+    public static class Entry {
+        /** The class that this entry can read. */
+        public final Class<?> categoryClass;
+
+        /** A name for the entry which is unique within the {@link #categoryClass}. */
+        public final ParseField name;
+
+        /** A parser for which returns subclasses of {@link #categoryClass}. */
+        public final FromXContent<?, ?> fromXContent;
+
+        /** Creates a new entry which can be stored by the registry. */
+        public <T> Entry(Class<T> categoryClass, ParseField name, FromXContent<? extends T, ?> fromXContent) {
+            this.categoryClass = Objects.requireNonNull(categoryClass);
+            this.name = Objects.requireNonNull(name);
+            this.fromXContent = Objects.requireNonNull(fromXContent);
+        }
+    }
+    
+    private final Map<Class<?>, Map<String, Entry>> registry;
+
+    /**
+     * Build the registry from a list of the entries that should be in it.
+     */
+    public NamedXContentRegistry(List<Entry> entries) {
+        Map<Class<?>, Map<String, Entry>> registry = new HashMap<>();
+        for (Entry entry : entries) {
+            Map<String, Entry> parsers = registry.get(entry.categoryClass);
+            if (parsers == null) {
+                parsers = new HashMap<>();
+                registry.put(entry.categoryClass, parsers);
+            }
+            for (String name : entry.name.getAllNamesIncludedDeprecated()) {
+                Object old = parsers.putIfAbsent(name, entry);
+                if (old != null) {
+                    throw new IllegalArgumentException(
+                            "Duplicate named XContent parsers for [" + entry.categoryClass.getName() + "][" + entry.name + "]");
+                }
+            }
+        }
+        registry.replaceAll((categoryClass, map) -> unmodifiableMap(map));
+        this.registry = unmodifiableMap(registry);
+    }
+
+    /**
+     * Wrap an {@link XContentParser} in one that implements {@link XContentParser#namedObject(Class, String, ParseFieldMatcherSupplier)}
+     * against this registry.
+     */
+    public XContentParser wrap(XContentParser parser) {
```

Comment:
Why isn't this inside xcontentfactory.creatParser? In fact why do we need a wrapper at all? Can't this just be native methods on XContentParser?